handler/par/flow_pushed_authorize.go

```diff
@@ -0,0 +1,88 @@
+package par
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/ory/fosite"
+	"github.com/ory/fosite/token/hmac"
+	"github.com/ory/x/errorsx"
+)
+
+const (
+	defaultPARKeyLength = 32
+)
+
+var b64 = base64.URLEncoding.WithPadding(base64.NoPadding)
+
+// PushedAuthorizeHandler handles the PAR request
+type PushedAuthorizeHandler struct {
+	Storage                  interface{}
+	PARContextLifetime       time.Duration
+	RequestURIPrefix         string
+	ScopeStrategy            fosite.ScopeStrategy
+	AudienceMatchingStrategy fosite.AudienceMatchingStrategy
+
+	IsRedirectURISecure func(*url.URL) bool
+}
+
+// HandlePushedAuthorizeEndpointRequest handles a pushed authorize endpoint request. To extend the handler's capabilities, the http request
+// is passed along, if further information retrieval is required. If the handler feels that he is not responsible for
+// the pushed authorize request, he must return nil and NOT modify session nor responder neither requester.
+func (c *PushedAuthorizeHandler) HandlePushedAuthorizeEndpointRequest(ctx context.Context, ar fosite.AuthorizeRequester, resp fosite.PushedAuthorizeResponder) error {
+	storage, ok := c.Storage.(fosite.PARStorage)
+	if !ok {
+		return errorsx.WithStack(fosite.ErrServerError.WithHint("Invalid storage type"))
```

Comment:
Could we return a better error here - such as `OAuth2 storage provider does not support Pushed Authorization Requests`?

Reply:
Yes, we can do that. The expectation would be to never hit this state anyway. This would be a major misconfiguration, where the service is writing a HTTP handler calling into the PAR handlers (NewPushAuthorizationRequest etc.) without actually implementing the storage interface. This was really more of a safety net.

Reply:
Makes sense! Good error messages can help tremendously when debugging, so even if this a coding error I think it just gives that extra 1% better developer experience :)

Reply:
Done